### ISSUE TITLE
Query with OR predicates with a CAST perform slower than IN list equivalent

### ISSUE THREAD

**Issue by akshatnair** [prestodb/presto#9133]
If there is a CAST present in the predicate then OR version of the query is very slow compared to the equivalent IN list query. There seems to be extra CAST operators that can be removed for optimization.

To reproduce this issue, create a view in memory connector based on tpch.sf10.orders table.

`
create view v1 as select orderkey, cast(comment as char(100)) comment from tpch.sf10.orders;
`
Explain analyze output for IN list query :

`explain analyze select 1 from v1 where comment in ('8495600010986943', '8495600010990366', '8495600010996678', '8495600011020536', '8495600011030741', '8495600011037027', '8495600011063460', '8495600011097104', '8495600020235695', '8495600020436350', '8495600020443786', '8495600020446805', '8495600031373022', '8495600031377213', '8495600031381017', '8495600031410725', '8495600031462460', '8495600072528765', '8495600072692314', '8495600072769229', '8495600072833090', '8495600073201578', '8495600073234181', '8495600073238174', '8495600073287825', '8495600073379234', '8495600073387138', '8495600073474597', '8495600073477889', '8495600073482996', '8495600073484711', '8495600073515605', '8495600073519631', '8495600073523773', '8495600073525018', '8495600073548275', '8495600073560015', '8495600073570436', '8495600073588941', '8495600073605893');`

```
Fragment 1 [tpch:orders:15000000]
    CPU: 39.58s, Input: 0 rows (0B); per task: avg.: 0.00 std.dev.: 0.00, Output: 0 rows (0B)
    Output layout: [expr_9]
    Output partitioning: SINGLE []
    - Project[] => [expr_9:integer]
            CPU fraction: 0.00%, Input: 0 rows (0B), Output: 0 rows (0B), Filtered: ?%
            Input avg.: 0.00 rows, Input std.dev.: ?%
            expr_9 := 1
        - ScanFilterProject[table = tpch:tpch:orders:sf10.0, originalConstraint = (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) IN (CAST('8495600010986943' AS varchar(100)), CAST('8495600010990366' AS varchar(100)), CAST('8495600010996678' AS varchar(100)), CAST('8495600011020536' AS varchar(100)), CAST('8495600011030741' AS varchar(100)), CAST('8495600011037027' AS varchar(100)), CAST('8495600011063460' AS varchar(100)), CAST('8495600011097104' AS varchar(100)), CAST('8495600020235695' AS varchar(100)), CAST('8495600020436350' AS varchar(100)), CAST('8495600020443786' AS varchar(100)), CAST('8495600020446805' AS varchar(100)), CAST('8495600031373022' AS varchar(100)), CAST('8495600031377213' AS varchar(100)), CAST('8495600031381017' AS varchar(100)), CAST('8495600031410725' AS varchar(100)), CAST('8495600031462460' AS varchar(100)), CAST('8495600072528765' AS varchar(100)), CAST('8495600072692314' AS varchar(100)), CAST('8495600072769229' AS varchar(100)), CAST('8495600072833090' AS varchar(100)), CAST('8495600073201578' AS varchar(100)), CAST('8495600073234181' AS varchar(100)), CAST('8495600073238174' AS varchar(100)), CAST('8495600073287825' AS varchar(100)), CAST('8495600073379234' AS varchar(100)), CAST('8495600073387138' AS varchar(100)), CAST('8495600073474597' AS varchar(100)), CAST('8495600073477889' AS varchar(100)), CAST('8495600073482996' AS varchar(100)), CAST('8495600073484711' AS varchar(100)), CAST('8495600073515605' AS varchar(100)), CAST('8495600073519631' AS varchar(100)), CAST('8495600073523773' AS varchar(100)), CAST('8495600073525018' AS varchar(100)), CAST('8495600073548275' AS varchar(100)), CAST('8495600073560015' AS varchar(100)), CAST('8495600073570436' AS varchar(100)), CAST('8495600073588941' AS varchar(100)), CAST('8495600073605893' AS varchar(100)))), 
filterPredicate = (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) IN (CAST('8495600010986943' AS varchar(100)), CAST('8495600010990366' AS varchar(100)), CAST('8495600010996678' AS varchar(100)), CAST('8495600011020536' AS varchar(100)), CAST('8495600011030741' AS varchar(100)), CAST('8495600011037027' AS varchar(100)), CAST('8495600011063460' AS varchar(100)), CAST('8495600011097104' AS varchar(100)), CAST('8495600020235695' AS varchar(100)), CAST('8495600020436350' AS varchar(100)), CAST('8495600020443786' AS varchar(100)), CAST('8495600020446805' AS varchar(100)), CAST('8495600031373022' AS varchar(100)), CAST('8495600031377213' AS varchar(100)), CAST('8495600031381017' AS varchar(100)), CAST('8495600031410725' AS varchar(100)), CAST('8495600031462460' AS varchar(100)), CAST('8495600072528765' AS varchar(100)), CAST('8495600072692314' AS varchar(100)), CAST('8495600072769229' AS varchar(100)), CAST('8495600072833090' AS varchar(100)), CAST('8495600073201578' AS varchar(100)), CAST('8495600073234181' AS varchar(100)), CAST('8495600073238174' AS varchar(100)), CAST('8495600073287825' AS varchar(100)), CAST('8495600073379234' AS varchar(100)), CAST('8495600073387138' AS varchar(100)), CAST('8495600073474597' AS varchar(100)), CAST('8495600073477889' AS varchar(100)), CAST('8495600073482996' AS varchar(100)), CAST('8495600073484711' AS varchar(100)), CAST('8495600073515605' AS varchar(100)), CAST('8495600073519631' AS varchar(100)), CAST('8495600073523773' AS varchar(100)), CAST('8495600073525018' AS varchar(100)), CAST('8495600073548275' AS varchar(100)), CAST('8495600073560015' AS varchar(100)), CAST('8495600073570436' AS varchar(100)), CAST('8495600073588941' AS varchar(100)), CAST('8495600073605893' AS varchar(100))))] => []
                CPU fraction: 100.00%, Input: 15000000 rows (0B), Output: 0 rows (0B), Filtered: 100.00%
                Input avg.: 3750000.00 rows, Input std.dev.: 0.00%
                comment := tpch:comment

```

OR list query:

`explain analyze select 1 from v1 where (comment = '8495600010986943' or comment = '8495600010990366' or comment = '8495600010996678' or comment = '8495600011020536' or comment = '8495600011030741' or comment = '8495600011037027' or comment = '8495600011063460' or comment = '8495600011097104' or comment = '8495600020235695' or comment = '8495600020436350' or comment = '8495600020443786' or comment = '8495600020446805' or comment = '8495600031373022' or comment = '8495600031377213' or comment = '8495600031381017' or comment = '8495600031410725' or comment = '8495600031462460' or comment = '8495600072528765' or comment = '8495600072692314' or comment = '8495600072769229' or comment = '8495600072833090' or comment = '8495600073201578' or comment = '8495600073234181' or comment = '8495600073238174' or comment = '8495600073287825' or comment = '8495600073379234' or comment = '8495600073387138' or comment = '8495600073474597' or comment = '8495600073477889' or comment = '8495600073482996' or comment = '8495600073484711' or comment = '8495600073515605' or comment = '8495600073519631' or comment = '8495600073523773' or comment = '8495600073525018' or comment = '8495600073548275' or comment = '8495600073560015' or comment = '8495600073570436' or comment = '8495600073588941' or comment = '8495600073605893' );`

```
Fragment 1 [tpch:orders:15000000]
    CPU: 4.08m, Input: 0 rows (0B); per task: avg.: 0.00 std.dev.: 0.00, Output: 0 rows (0B)
    Output layout: [expr_9]
    Output partitioning: SINGLE []
    - Project[] => [expr_9:integer]
            CPU fraction: 0.00%, Input: 0 rows (0B), Output: 0 rows (0B), Filtered: ?%
            Input avg.: 0.00 rows, Input std.dev.: ?%
            expr_9 := 1
        - ScanFilterProject[table = tpch:tpch:orders:sf10.0, originalConstraint = (((((((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600010986943' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600010990366' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600010996678' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600011020536' AS varchar(100))))) OR (((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600011030741' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600011037027' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600011063460' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600011097104' AS varchar(100)))))) OR ((((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600020235695' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600020436350' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600020443786' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600020446805' AS varchar(100))))) OR (((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600031373022' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600031377213' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600031381017' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600031410725' AS varchar(100))))))) OR (((((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600031462460' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600072528765' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600072692314' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600072769229' AS varchar(100))))) OR (((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600072833090' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073201578' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073234181' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073238174' AS varchar(100)))))) OR ((((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073287825' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073379234' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073387138' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073474597' AS varchar(100))))) OR (((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073477889' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073482996' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073484711' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073515605' AS varchar(100)))))))) OR ((((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073519631' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073523773' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073525018' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073548275' AS varchar(100))))) OR (((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073560015' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073570436' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073588941' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073605893' AS varchar(100))))))), filterPredicate = (((((((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600010986943' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600010990366' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600010996678' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600011020536' AS varchar(100))))) OR (((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600011030741' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600011037027' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600011063460' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600011097104' AS varchar(100)))))) OR ((((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600020235695' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600020436350' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600020443786' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600020446805' AS varchar(100))))) OR (((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600031373022' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600031377213' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600031381017' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600031410725' AS varchar(100))))))) OR (((((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600031462460' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600072528765' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600072692314' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600072769229' AS varchar(100))))) OR (((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600072833090' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073201578' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073234181' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073238174' AS varchar(100)))))) OR ((((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073287825' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073379234' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073387138' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073474597' AS varchar(100))))) OR (((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073477889' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073482996' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073484711' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073515605' AS varchar(100)))))))) OR ((((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073519631' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073523773' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073525018' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073548275' AS varchar(100))))) OR (((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073560015' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073570436' AS varchar(100)))) OR ((CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073588941' AS varchar(100))) OR (CAST(CAST(CAST(""comment"" AS char(100)) AS char(100)) AS varchar(100)) = CAST('8495600073605893' AS varchar(100)))))))] => []
                CPU fraction: 100.00%, Input: 15000000 rows (0B), Output: 0 rows (0B), Filtered: 100.00%
                Input avg.: 3750000.00 rows, Input std.dev.: 0.00%
                comment := tpch:comment
```

**Comment by akshatnair** [comment]
@martint @dain any thoughts on this? 
We may have to add a way to either remove these extra CASTs or merge the ORs as in list to avoid this casting multiple times for column names.

**Comment by akshatnair** [comment]
This problem has been partially fixed by the fix for #9136. We will still have a scenario when an underlying table with `char(n)` column is used for comparisons. It will get CAST to `varchar(n)` for comparisons and that will make OR slower than IN.

One possible solution is to combine such ORs to IN if we have a function in that expression on a column.

**Comment by findepi** [comment]
> One possible solution is to combine such `OR`s to `IN`

This would also play nicely with plan node statistics derivation. We have a rule for `IN`  predicate (just as for `=`, `<=` etc.) whereas generic rule for `OR` is less precise.
